File: src/finance/trading/factor/proj_grad.py
from dataclasses import dataclass
import numpy as np
from numpy.linalg import norm
import pandas as pd
from typing import Tuple, Optional
from scipy.linalg import svd, inv, sqrtm
from numba import guvectorize, float64, int64
from tqdm.auto import tqdm
from finance.trading.factor.__model import Model
from finance.trading.factor.benchmark import Benchmark


@dataclass
class ProjGrad(Model):
    step_stiefel: float = 0.1
    step_beta: float = 0.05
    random_state: int = 1234
    n_iter: int = 1000
    use_benchmark: bool = False

    def __post_init__(self):
        if self.use_benchmark:
            self.__benchmark = Benchmark(
                lag=self.lag,
                n_factors=self.n_factors,
                eps=self.eps,
                n_iter=self.n_iter,
                random_state=self.random_state,
            )

    def train(
        self,
        X_train: np.ndarray,
        Y_train: np.ndarray,
        stiefel0: Optional[np.ndarray] = None,
        beta0: Optional[np.ndarray] = None,
    ) -> Tuple[np.ndarray, np.ndarray]:
        if self.use_benchmark:
            stiefel0, beta0 = self.__benchmark.train(X_train, Y_train)
        elif stiefel0 is None or beta0 is None:
            raise ValueError(
                "Initial values should be provided for stiefel0 and beta0."
            )

        stiefel = stiefel0.copy()
        beta = beta0.copy()
        stiefel_star = stiefel.copy()
        beta_star = beta.copy()
        if len(beta.shape) == 1:
            beta = beta.reshape(-1, 1)
        max_metric = self.metric_train(stiefel, beta, X_train, Y_train)
        identity_lag = np.eye(self.lag)
        identity_factors = np.eye(self.n_factors)
        factors = np.arange(self.n_factors)
        n_assets = Y_train.shape[0]
        n_dates = X_train.shape[0] // n_assets
        gen = np.random.default_rng(self.random_state)
        iterations = gen.choice(np.arange(n_dates), self.n_iter, replace=True)
        for index in tqdm(iterations, desc=self.name):
            return_date = Y_train[:, index]
            # print(f"return_date shape: {return_date.shape}")
            # print(f"\n{date} return_date: \n{return_date}")
            # lagged_returns = X_train.loc[self.get_index(date, assets), :].to_numpy()
            start = index * n_assets
            end = (index + 1) * n_assets
            lagged_returns = X_train[start:end, :]
            # print(f"lagged_returns shape: {lagged_returns.shape}")
            # print(f"\n{date} lagged_returns: \n{lagged_returns}")
            prediction = lagged_returns @ stiefel @ beta.reshape(-1, 1)
            # print(f"prediction shape: {prediction.shape}")
            # print(f"prediction.T shape: {prediction.T.shape}")
            # print(f"\n{date} prediction: \n{prediction}")
            returns = return_date.T @ lagged_returns / np.sqrt(norm(return_date, 2))
            # print(f"returns shape: {returns.shape}")
            # print(f"\n{date} returns: \n{returns}")

            # gradients
            grad_stiefel = self.grad_stiefel(
                returns,
                beta.reshape(-1, 1),
                factors,
                stiefel,
                prediction.reshape(1, -1),
                lagged_returns,
                identity_lag,
            ).T.reshape(self.lag, self.n_factors)
            # print(f"\n{date} grad_stiefel: \n{grad_stiefel}")
            grad_beta = self.grad_beta(
                returns,
                beta,
                stiefel,
                prediction,
                lagged_returns,
                identity_factors,
            )
            # print(f"\n{date} grad_beta: \n{grad_beta}")

            # updates
            beta = beta + self.step_beta * grad_beta
            # beta = fitBeta(A)
            # print(f"\n{date} beta: \n{beta}")
            stiefel_temp = stiefel + self.step_stiefel * grad_stiefel
            stiefel = self.project(stiefel_temp)
            # stiefel = stiefel_temp @ inv(sqrtm(stiefel_temp.T @ stiefel_temp))
            # print(f"\n{date} stiefel: \n{stiefel}")

            m = self.metric_train(stiefel, beta, X_train, Y_train)

            if m > max_metric:
                tqdm.write(f"Iteration: {index+1} with best train metric: {m}")
                max_metric = m
                stiefel_star = stiefel
                beta_star = beta

        return stiefel_star, beta_star

    @staticmethod
    @guvectorize(
        [
            (
                float64[:],
                float64[:, :],
                int64,
                float64[:, :],
                float64[:, :],
                float64[:, :],
                float64[:, :],
                float64[:, :],
            )
        ],
        "(q),(m,k),(),(p,m),(k,n),(n,q),(q,q)->(k,q)",
        nopython=True,
    )
    def grad_stiefel(
        returns: np.ndarray,
        beta: np.ndarray,
        factor: int,
        stiefel: np.ndarray,
        prediction: np.ndarray,
        lagged_returns: np.ndarray,
        identity_lag: np.ndarray,
        res: np.ndarray,
    ):
        res[:, :] = (
            returns
            @ (
                beta[factor][0] * identity_lag
                - beta[factor][0]
                * stiefel
                @ beta
                @ prediction
                @ lagged_returns
                / np.sqrt(np.linalg.norm(prediction, 2)) ** 3
            )
        ).reshape(1, -1)[:, :]

    @staticmethod
    def grad_beta(
        returns: np.ndarray,
        beta: np.ndarray,
        stiefel: np.ndarray,
        prediction: np.ndarray,
        lagged_returns: np.ndarray,
        identity_factors: np.ndarray,
    ):
        return (
            returns
            @ stiefel
            @ (
                identity_factors
                - beta.reshape(-1, 1)
                @ prediction.reshape(1, -1)
                @ lagged_returns
                @ stiefel
                / np.sqrt(np.linalg.norm(prediction, 2)) ** 3
            )
        ).reshape(-1, 1)

    @staticmethod
    def project(stiefel: np.ndarray) -> np.ndarray:
        u, _, v = svd(stiefel, full_matrices=False)
        return u @ v

File: rust/finance/src/trading/factor/proj_grad.rs
#[cfg(test)]
mod test {
    use crate::loader::Loader;
    use crate::trading::factor::benchmark::BenchMark;
    use crate::trading::factor::preprocessor::{DataType, Preprocessor};
    use crate::trading::factor::utils::{calc_metric, norm_matrix, norm_vec, svd, MatrixType};
    use ndarray::parallel::prelude::*;
    use ndarray::{concatenate, s, Array, Axis};
    use ndarray_rand::rand_distr::Uniform;
    use ndarray_rand::{rand::SeedableRng, RandomExt};
    use polars::prelude::ScanArgsParquet;
    use rand_chacha::ChaCha20Rng;
    use std::time::{Duration, Instant};
    #[test]
    fn test_code() {
        let start = Instant::now();
        let path = "../../data/returns.parquet";
        let loader = Loader::init(path);
        let data = loader.load_parquet(ScanArgsParquet::default());
        let max_lag = 250;
        let preprocessor = Preprocessor::init(max_lag);
        let (x_train, y_train) = preprocessor.split_x_y_ndarray(data, DataType::RETURN, 1);
        let duration = start.elapsed();
        println!("Time elapsed data processing is: {:?}", duration);
        // train(&x_train, &y_train);
        let eps = 1e-6;
        let max_lag = 250;
        let n_factors = 3;
        let n_iter = 150;
        let tol = 1e-6;
        let random_state = 12345;
        let stiefel_step = 0.1;
        let beta_step = 0.05;
        let start = Instant::now();
        let mut benchmark_model = BenchMark::init(100, tol, random_state, max_lag, n_factors);
        let (stiefel, beta) = benchmark_model.parallel_train(&x_train, &y_train, Some(5));
        // let duration = start.elapsed();
        // println!("Time elapsed parallel training: {:?}", duration);
        let start = Instant::now();
        // let mut stiefel = Array::ones((max_lag, n_factors));
        // let mut beta = Array::ones((n_factors, 1)); // .reshape((-1, 1));
        //                                             // println!("stiefel: \n{:?}", stiefel);
        //                                             // println!("beta: \n{:?}", beta);
        //                                             // let (stiefel, beta) = benchmark_model.sequential_train(&x_train, &y_train);
        //                                             // let duration = start.elapsed();
        //                                             // println!("Time elapsed sequential training: {:?}", duration);
        let mut stiefel_star = Array::zeros((max_lag, n_factors));
        let mut beta_star = Array::zeros((n_factors, 1));
        let mut max_metric = calc_metric(eps, &stiefel, &beta, &x_train, &y_train);
        let identity_lag: MatrixType<f64> = Array::eye(max_lag);
        let identity_factors: MatrixType<f64> = Array::eye(n_factors);
        let n_assets = y_train.shape()[0];
        let n_dates = x_train.shape()[0] / n_assets;
        let mut rng = ChaCha20Rng::seed_from_u64(random_state);
        let random_dates =
            Array::random_using(n_iter, Uniform::new_inclusive(0, n_dates - 1), &mut rng);
        let factors: Vec<usize> = (0..n_factors).collect();
        for (pos, date) in random_dates.iter().enumerate() {
            if pos > 10 {
                break;
            }
            // println!("xtrain: \n{:?}", x_train);
            // for date in &[493] {
            let return_date = y_train.slice(s![.., *date]);
            // println!("return_date: \n{:?}\n", return_date);
            let start = *date * n_assets;
            let end = (*date + 1) * n_assets;
            let lagged_returns = x_train.slice(s![start..end, ..]);
            // println!("lagged_returns: \n{:?}\n", lagged_returns);
            let prediction = lagged_returns.dot(&stiefel).dot(&beta);
            // println!("prediction: \n{:?}\n", prediction);
            // println!("{date}");
            // println!("{:?}", stiefel.shape());
            let temp_ret = return_date.t().dot(&lagged_returns);
            // println!("temp_ret: \n{:?}\n", temp_ret);
            let norm = norm_vec(return_date).sqrt();
            // println!("norm: \n{:?}\n", norm);
            let returns = temp_ret / norm;
            // println!("returns: \n{:?}\n", returns);

            // println!("{:?}", returns);
            // Gradients.
            let grad_stiefel = factors
                .iter()
                .map(|factor| {
                    let derivative = returns.dot(
                        &(beta[[*factor, 0]] * &identity_lag
                            - beta[[*factor, 0]]
                                * &stiefel.dot(&beta).dot(&prediction.t()).dot(&lagged_returns)
                                / norm_matrix(&prediction).sqrt().powi(3)),
                    );
                    // println!("ZERO:\n{:?}", derivative);
                    let len = derivative.shape()[0];
                    let derivative = derivative.into_shape((len, 1)).unwrap();
                    // println!("UN:\n{:?}", derivative);
                    derivative
                })
                .collect::<Vec<MatrixType<f64>>>();
            // println!("\ngrad 0: \n{:?}", &grad_stiefel[0]);
            let grad_stiefel: Vec<_> = grad_stiefel
                .iter()
                .map(|ret| ret.slice(s![.., ..]))
                .collect();
            let grad_stiefel =
                concatenate(Axis(1), &grad_stiefel).expect("Failed to get grad Stiefel");
            // println!("grad stiefel:\n{:?}", grad_stiefel);

            let grad_beta = returns.dot(&stiefel).dot(
                &(&identity_factors
                    - &beta.dot(&prediction.t()).dot(&lagged_returns).dot(&stiefel)
                        / norm_matrix(&prediction).sqrt().powi(3)),
            );
            let len = grad_beta.shape()[0];
            let grad_beta = grad_beta.into_shape((len, 1)).unwrap();
            // println!("grad beta:\n{:?}", grad_beta);
            // println!("\n{:?}", beta.shape());
            // println!("\n{:?}", grad_beta.shape());
            // Updates.
            let beta = &beta + beta_step * &grad_beta;
            let mut stiefel = &stiefel + stiefel_step * &grad_stiefel;
            let (u, s, v) = svd(&stiefel, None, 1000, 123);
            // println!("u: \n{:?}", u);
            // println!("s: \n{:?}", s);
            // println!("v: \n{:?}", v);
            stiefel = u.dot(&v);

            // println!("{:?}", beta.shape());
            // println!("{:?}", stiefel.shape());
            let metric = calc_metric(eps, &stiefel, &beta, &x_train, &y_train);
            println!("{metric}");
            if metric > max_metric {
                println!("Best metric: {metric}");
                stiefel_star = stiefel;
                beta_star = beta;
                max_metric = metric;
            }
        }
    }

    fn train(x_train: &MatrixType<f64>, y_train: &MatrixType<f64>) {
        let eps = 1e-6;
        let max_lag = 250;
        let n_factors = 10;
        let n_iter = 150;
        let tol = 1e-6;
        let random_state = 12345;
        let stiefel_step = 0.1;
        let beta_step = 0.05;
        let mut benchmark_model = BenchMark::init(40, tol, random_state, max_lag, n_factors);
        let (stiefel, beta) = benchmark_model.parallel_train(x_train, y_train, Some(5));
        // let (stiefel, beta) = benchmark_model.sequential_train(x_train, y_train);
        // let mut stiefel_star = Array::zeros((max_lag, n_factors));
        // let mut beta_star = Array::zeros((n_factors, 1));
        // let mut max_metric = calc_metric(eps, &stiefel, &beta, x_train, y_train);
        // let identity_lag: MatrixType<f64> = Array::eye(max_lag);
        // let identity_factors: MatrixType<f64> = Array::eye(n_factors);
        // let n_assets = y_train.shape()[0];
        // let n_dates = x_train.shape()[0] / n_assets;
        // let mut rng = ChaCha20Rng::seed_from_u64(random_state);
        // let random_dates =
        //     Array::random_using(n_iter, Uniform::new_inclusive(0, n_dates - 1), &mut rng);
        // let factors: Vec<usize> = (0..n_factors).collect();
        // for date in &random_dates {
        //     let return_date = y_train.slice(s![.., *date]);
        //     let start = *date * n_assets;
        //     let end = (*date + 1) * n_assets;
        //     let lagged_returns = x_train.slice(s![start..end, ..]);
        //     let prediction = lagged_returns.dot(&stiefel).dot(&beta);
        //     // println!("{date}");
        //     // println!("{:?}", stiefel.shape());
        //     let returns = return_date.t().dot(&lagged_returns) / norm_vec(return_date);

        //     // println!("{:?}", returns);
        //     // Gradients.
        //     let grad_stiefel = factors
        //         .iter()
        //         .map(|factor| {
        //             let derivative = returns.dot(
        //                 &(beta[[*factor, 0]] * &identity_lag
        //                     - beta[[*factor, 0]]
        //                         * &stiefel.dot(&beta).dot(&prediction.t()).dot(&lagged_returns)),
        //             ) / norm_matrix(&prediction).powi(3);
        //             // println!("ZERO:\n{:?}", derivative);
        //             let len = derivative.shape()[0];
        //             let derivative = derivative.into_shape((len, 1)).unwrap();
        //             // println!("UN:\n{:?}", derivative);
        //             derivative
        //         })
        //         .collect::<Vec<MatrixType<f64>>>();
        //     let grad_stiefel: Vec<_> = grad_stiefel
        //         .iter()
        //         .map(|ret| ret.slice(s![.., ..]))
        //         .collect();
        //     let grad_stiefel =
        //         concatenate(Axis(1), &grad_stiefel).expect("Failed to get grad Stiefel");

        //     let grad_beta = returns.dot(&stiefel).dot(
        //         &(&identity_factors
        //             - &beta.dot(&prediction.t()).dot(&lagged_returns).dot(&stiefel)),
        //     ) / norm_matrix(&prediction).powi(3);
        //     let len = grad_beta.shape()[0];
        //     let grad_beta = grad_beta.into_shape((len, 1)).unwrap();
        //     // println!("\n{:?}", beta.shape());
        //     // println!("\n{:?}", grad_beta.shape());
        //     // Updates.
        //     let beta = &beta + beta_step * &grad_beta;
        //     let stiefel = &stiefel + stiefel_step * &grad_stiefel;
        //     // println!("{:?}", beta.shape());

        //     // println!("{:?}", beta.shape());
        //     // println!("{:?}", stiefel.shape());
        //     let metric = calc_metric(eps, &stiefel, &beta, x_train, y_train);
        //     if metric > max_metric {
        //         println!("Best metric: {metric}");
        //         stiefel_star = stiefel;
        //         beta_star = beta;
        //         max_metric = metric;
        //     }
        // }
    }
}
